src/merkle/store.rs

```diff
@@ -0,0 +1,375 @@
+
+        let mut depth = 0;
+        let mut parent_offset = 1;
+        let mut child_offset = 2;
+        while depth < layers {
+            let layer_size = 1usize << depth;
+            for _ in 0..layer_size {
+                // merkle tree is using level form representation, so left and right siblings are
+                // next to each other
+                let left = tree.nodes[child_offset];
+                let right = tree.nodes[child_offset + 1];
+                self.nodes.insert(
+                    tree.nodes[parent_offset].into(),
+                    Node {
+                        left: left.into(),
+                        right: right.into(),
+                    },
+                );
+                parent_offset += 1;
+                child_offset += 2;
+            }
+            depth += 1;
+        }
+
+        Ok(tree.nodes[1])
+    }
+
+    pub fn add_sparse_merkle_tree<R, I>(&mut self, entries: R) -> Result<Word, MerkleError>
+    where
+        R: IntoIterator<IntoIter = I>,
+        I: Iterator<Item = (u64, Word)> + ExactSizeIterator,
+    {
+        let smt = SimpleSmt::new(SimpleSmt::MAX_DEPTH)?.with_leaves(entries)?;
+        for branch in smt.store.branches.values() {
+            let parent = Rpo256::merge(&[branch.left, branch.right]);
+            self.nodes.insert(
+                parent.into(),
+                Node {
+                    left: branch.left.into(),
+                    right: branch.right.into(),
+                },
+            );
+        }
+
+        Ok(smt.root())
+    }
+
+    pub fn add_merkle_paths<I>(&mut self, paths: I) -> Result<Word, MerkleError>
+    where
+        I: IntoIterator<Item = (u64, Word, MerklePath)>,
+    {
+        let node: Word = Word::default();
+
+        for (index_value, mut node, path) in paths {
+            let mut index = NodeIndex::new(self.nodes.len() as u8, index_value);
+
+            for sibling in path {
+                let (left, right) = match index.is_value_odd() {
+                    true => (sibling, node),
+                    false => (node, sibling),
+                };
+                let parent = Rpo256::merge(&[left.into(), right.into()]).into();
+                self.nodes.insert(
+                    parent,
+                    Node {
+                        left: left.into(),
+                        right: right.into(),
+                    },
+                );
+
+                index.move_up();
+                node = parent.word;
+            }
+        }
+
+        // Returns the parent of the last paths (assumes all paths have the same parent) or empty
+        Ok(node)
+    }
+
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    pub fn get_node(&self, root: Word, index: NodeIndex) -> Result<Word, MerkleError> {
+        let mut hash: WordWithOrd = root.into();
+        let directions = index.directions();
```

Comment:
I think we should just `move_up` instead of allocating a vec for every `get_node` call

Reply:
ref: https://github.com/0xPolygonMiden/crypto/pull/93#discussion_r1134840801